packages/js/components/src/experimental-select-control/async-select-control.tsx

```diff
@@ -0,0 +1,168 @@
+/**
+ * External dependencies
+ */
+import { Spinner } from '@wordpress/components';
+import {
+	createElement,
+	useState,
+	useMemo,
+	useCallback,
+	useEffect,
+} from '@wordpress/element';
+import { debounce } from 'lodash';
+import { ReactElement, Component } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { SelectControlProps, SelectControl } from './select-control';
+import { Menu } from './menu';
+import { MenuItem } from './menu-item';
+import {
+	defaultGetFilteredItems,
+	defaultGetItemLabel,
+	defaultGetItemValue,
+} from './utils';
+import { ChildrenProps } from './types';
+
+type AsyncSelectControlProps< ItemType > = Omit<
+	SelectControlProps< ItemType >,
+	'onInputChange' | 'items' | 'children'
+> & {
+	onSearch: ( value: string | undefined ) => Promise< ItemType[] >;
+	items?: ItemType[];
+	triggerInitialFetch?: boolean;
+	pageSize?: number;
+	total?: number;
+	children: (
+		props: ChildrenProps< ItemType > & {
+			isFetching: boolean;
+		}
+	) => ReactElement | Component;
+};
+
+export function AsyncSelectControl< ItemType >( {
+	onSearch,
+	getItemLabel = defaultGetItemLabel,
+	getItemValue = defaultGetItemValue,
+	items = [],
+	triggerInitialFetch = true,
+	pageSize,
+	total,
+	children = ( {
+		items: renderItems,
+		highlightedIndex,
+		getItemProps,
+		getMenuProps,
+		isOpen,
+		isFetching,
+	} ) => {
+		return (
+			<Menu isOpen={ isOpen } getMenuProps={ getMenuProps }>
+				{ isFetching ? (
+					<Spinner />
+				) : (
+					renderItems.map( ( item, index: number ) => (
+						<MenuItem
+							key={ `${ getItemValue( item ) }` }
+							index={ index }
+							isActive={ highlightedIndex === index }
+							item={ item }
+							getItemProps={ getItemProps }
+						>
+							{ getItemLabel( item ) }
+						</MenuItem>
+					) )
+				) }
+			</Menu>
+		);
+	},
+	...remainingSelectControlProps
+}: AsyncSelectControlProps< ItemType > ) {
+	const [ fetchedItems, setFetchedItems ] = useState< ItemType[] >( items );
+	const [ isFetching, setIsFetching ] = useState( false );
+
+	const fetchItems = useCallback(
+		( value: string | undefined ) => {
+			setIsFetching( true );
+			onSearch( value ).then(
+				( results ) => {
+					setFetchedItems( results );
+					setIsFetching( false );
+				},
+				() => {
+					setIsFetching( false );
+				}
+			);
+		},
+		[ setFetchedItems, setIsFetching, onSearch ]
+	);
+
+	useEffect( () => {
+		if (
+			triggerInitialFetch &&
+			! remainingSelectControlProps.disabled &&
+			items.length === 0
+		) {
+			fetchItems( '' );
+		}
+	}, [ remainingSelectControlProps.disabled ] );
+
+	const searchDebounced = useMemo(
```

Comment:
`useMemo` and `useCallback` do not guarantee that the functions won't be recreated, even if the dependencies do not change (I hit this in https://github.com/woocommerce/woocommerce/pull/34963). I think we might want to go with a similar solution as in #34963.

Reply:
Might want to consider `useDebounce` (see #34963)

Reply:
Otherwise, I think that this will suffer similar issues to what I found in DateTimePickerControl. But, since you use `debounce` instead of `useDebounce`, this might actually be "more" okay, as your debounce won't get canceled.